src/gui/vector/qgsattributesformmodel.cpp

```diff
@@ -0,0 +1,1162 @@
+      return true;
+    }
+    case QgsAttributesFormModel::NodeIdRole:
+    {
+      mNodeId = value.toString();
+      return true;
+    }
+    default:
+      return false;
+  }
+  return false;
+}
+
+
+//
+// QgsAttributesFormModel implementation
+//
+
+QgsAttributesFormModel::QgsAttributesFormModel( QgsVectorLayer *layer, QObject *parent )
+  : QAbstractItemModel( parent )
+  , mRootNode( std::make_unique< QgsAttributesFormTreeNode >() )
+  , mLayer( layer )
+{
+}
+
+QgsAttributesFormModel::~QgsAttributesFormModel() = default;
+
+QgsAttributesFormTreeNode *QgsAttributesFormModel::nodeForIndex( const QModelIndex &index ) const
+{
+  if ( index.isValid() )
+  {
+    if ( auto *node = static_cast<QgsAttributesFormTreeNode *>( index.internalPointer() ) )
+      return node;
+  }
+  return mRootNode.get();
+}
+
+int QgsAttributesFormModel::rowCount( const QModelIndex &parent ) const
+{
+  if ( parent.isValid() && parent.column() > 0 )
+    return 0;
+
+  const QgsAttributesFormTreeNode *parentNode = nodeForIndex( parent );
+
+  return parentNode ? parentNode->childCount() : 0;
+}
+
+int QgsAttributesFormModel::columnCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent );
+  return 1;
+}
+
+QStringList QgsAttributesFormModel::mimeTypes() const
+{
+  return QStringList() << QStringLiteral( "application/x-qgsattributetabledesignerelement" );
+}
+
+QMimeData *QgsAttributesFormModel::mimeData( const QModelIndexList &indexes ) const
+{
+  if ( indexes.count() == 0 )
+    return nullptr;
+
+  const QStringList types = mimeTypes();
+  if ( types.isEmpty() )
+    return nullptr;
+
+  QMimeData *data = new QMimeData();
+  const QString format = types.at( 0 );
+  QByteArray encoded;
+  QDataStream stream( &encoded, QIODevice::WriteOnly );
+
+  // Sort indexes since their order reflects selection order
+  QModelIndexList sortedIndexes = indexes;
+  std::sort( sortedIndexes.begin(), sortedIndexes.end() );
+
+  for ( const QModelIndex &index : std::as_const( sortedIndexes ) )
+  {
+    if ( index.isValid() )
+    {
+      const QString nodeId = index.data( QgsAttributesFormModel::NodeIdRole ).toString();
+      const QString nodeName = index.data( QgsAttributesFormModel::NodeNameRole ).toString();
+      int nodeType = index.data( QgsAttributesFormModel::NodeTypeRole ).toInt();
+      const auto nodeData = index.data( QgsAttributesFormModel::NodeDataRole );
+
+      stream << nodeId << nodeType << nodeName << nodeData;
+    }
+  }
+
+  data->setData( format, encoded );
+  return data;
+}
+
+QModelIndex QgsAttributesFormModel::index( int row, int column, const QModelIndex &parent ) const
+{
+  if ( !hasIndex( row, column, parent ) )
+    return QModelIndex();
+
+  QgsAttributesFormTreeNode *parentNode = nodeForIndex( parent );
+  if ( !parentNode )
+    return QModelIndex();
+
+  if ( QgsAttributesFormTreeNode *childNode = parentNode->child( row ) )
+    return createIndex( row, column, childNode );
```

Comment:
isn't this supposed to be the parentNode here?

Reply:
[createIndex()](https://doc.qt.io/qt-6/qabstractitemmodel.html#createIndex) expects a pointer to the item for which we'd like the index.

Reply:
Right I mean why are we using childNode here and not parentNode?

Reply:
In that function we're using the `row` and `parent` params to access a specific child, because we're interested in that child's index. 

`parent` lets us access the underlying `parentItem`, from which we can get (using the `row`) the wanted `childItem`, which in turn we pass to `createIndex()` to get its correspondent model index.

Reply:
Is it clearer now?